File: cineva/packages/audio_engine/native/cineva_dsp/src/channel_mapper.c
#include "channel_mapper.h"

#include <string.h>

#include "dsp_util.h"

int cv_layout_from_channels(int channels) {
  switch (channels) {
    case 1: return 0;
    case 2: return 1;
    case 4: return 2;
    case 6: return 3;
    case 8: return 4;
    default: return channels > 2 ? 4 : (channels < 1 ? 0 : 1);
  }
}

void cv_mapper_weights(int layout, int channels, double *weights) {
  for (int i = 0; i < CINEVA_DSP_MAX_CHANNELS; i++) weights[i] = 1.0;
  if (layout >= 3 && channels >= 6) {
    weights[4] = 1.4142135623730951; /* Ls */
    weights[5] = 1.4142135623730951; /* Rs */
    weights[3] = 0.0;                /* LFE exclu de la mesure */
  }
  if (layout >= 4 && channels >= 8) {
    weights[6] = 1.4142135623730951; /* Lb */
    weights[7] = 1.4142135623730951; /* Rb */
  }
}

void cv_mapper_downmix_itu(const float *const *in, int channels, int layout,
                           float *const *out, double lfe_gain_lin, int frames) {
  double lg = cv_sanitize_d(lfe_gain_lin);
  if (lg < 0.0) lg = 0.0;

  if (channels <= 1) {
    for (int i = 0; i < frames; i++) {
      float x = in[0][i];
      out[0][i] = x;
      out[1][i] = x;
    }
    return;
  }

  int has_center = layout >= 3 && channels >= 6;
  int has_surround = layout >= 2 && channels >= 4;
  int has_lfe = layout >= 3 && channels >= 6;
  int has_back = layout >= 4 && channels >= 8;

  for (int i = 0; i < frames; i++) {
    double l = cv_sanitize_f(in[0][i]);
    double r = cv_sanitize_f(in[1][i]);
    if (has_center) {
      double c = cv_sanitize_f(in[2][i]) * 0.7071067811865476;
      l += c;
      r += c;
    }
    if (has_surround) {
      l += cv_sanitize_f(in[4 % channels][i]) * 0.7071067811865476;
      r += cv_sanitize_f(in[5 % channels][i]) * 0.7071067811865476;
    }
    if (has_back) {
      l += cv_sanitize_f(in[6][i]) * 0.7071067811865476;
      r += cv_sanitize_f(in[7][i]) * 0.7071067811865476;
    }
    if (has_lfe && lg > 0.0) {
      double lfe = cv_sanitize_f(in[3][i]) * lg;
      l += lfe;
      r += lfe;
    }
    out[0][i] = (float)l;
    out[1][i] = (float)r;
  }
}

File: cineva/packages/audio_engine/lib/src/dsp/channel_mapper.dart
/// Channel Mapper — miroir de `native/cineva_dsp/src/channel_mapper.c`.
library;

import 'biquad.dart' as dsp;

const int layoutMono = 0;
const int layoutStereo = 1;
const int layoutQuad = 2;
const int layout51 = 3;
const int layout71 = 4;

const int maxChannels = 8;

int layoutFromChannels(int channels) {
  switch (channels) {
    case 1:
      return layoutMono;
    case 2:
      return layoutStereo;
    case 4:
      return layoutQuad;
    case 6:
      return layout51;
    case 8:
      return layout71;
    default:
      return channels > 2 ? layout71 : (channels < 1 ? layoutMono : layoutStereo);
  }
}

/// Poids BS.1770 par canal (L/R/C = 1, surrounds = √2, LFE = 0).
void mapperWeights(int layout, int channels, List<double> weights) {
  for (int i = 0; i < maxChannels; i++) {
    weights[i] = 1.0;
  }
  if (layout >= layout51 && channels >= 6) {
    weights[4] = 1.4142135623730951; // Ls
    weights[5] = 1.4142135623730951; // Rs
    weights[3] = 0.0; // LFE exclu de la mesure
  }
  if (layout >= layout71 && channels >= 8) {
    weights[6] = 1.4142135623730951; // Lb
    weights[7] = 1.4142135623730951; // Rb
  }
}

/// Downmix ITU-R BS.775 vers la stéréo. `lfeGainLin` = 0 ⇒ LFE ignoré.
void downmixItu(
  List<Float32List> in_,
  int channels,
  int layout,
  List<Float32List> out,
  double lfeGainLin,
  int frames,
) {
  final double lg = dsp.sanitizeD(lfeGainLin).clamp(0.0, double.infinity);
  if (channels <= 1) {
    for (int i = 0; i < frames; i++) {
      out[0][i] = in_[0][i];
      out[1][i] = in_[0][i];
    }
    return;
  }
  final bool hasCenter = layout >= layout51 && channels >= 6;
  final bool hasSurround = layout >= layoutQuad && channels >= 4;
  final bool hasLfe = layout >= layout51 && channels >= 6;
  final bool hasBack = layout >= layout71 && channels >= 8;
  for (int i = 0; i < frames; i++) {
    double l = dsp.sanitizeD(in_[0][i]);
    double r = dsp.sanitizeD(in_[1][i]);
    if (hasCenter) {
      final double c = dsp.sanitizeD(in_[2][i]) * 0.7071067811865476;
      l += c;
      r += c;
    }
    if (hasSurround) {
      l += dsp.sanitizeD(in_[4 % channels][i]) * 0.7071067811865476;
      r += dsp.sanitizeD(in_[5 % channels][i]) * 0.7071067811865476;
    }
    if (hasBack) {
      l += dsp.sanitizeD(in_[6][i]) * 0.7071067811865476;
      r += dsp.sanitizeD(in_[7][i]) * 0.7071067811865476;
    }
    if (hasLfe && lg > 0) {
      final double lfe = dsp.sanitizeD(in_[3][i]) * lg;
      l += lfe;
      r += lfe;
    }
    out[0][i] = l;
    out[1][i] = r;
  }
}
